### ISSUE TITLE
Slurm detection in 3.3

### ISSUE THREAD

**Issue by mcuma** [pmodels/mpich#3572]
When running MPICH 3.3 on SLURM, the hostlist is not brought in from the SLURM. This is a different behavior from SLURM 3.2.1. To reproduce:

- build MPICH 3.2.1 and 3.3, for comparison with the same ch3 device (and a bit dated libfabric to be on the same page with 3.2.1):

```
../../../srcdir/mpich/3.2.1/configure \
    --prefix=/uufs/chpc.utah.edu/sys/installdir/mpich/3.2.1test --enable-romio \
    --with-file-system=nfs+ufs --with-mpe -with-device=ch3:nemesis:tcp,ofi \
    --enable-threads=runtime --enable-fast=all \
    --with-ofi=/uufs/chpc.utah.edu/sys/installdir/libfabric/1.6.0
```

```
../../../srcdir/mpich/3.3/configure \
    --prefix=/uufs/chpc.utah.edu/sys/installdir/mpich/3.3test --enable-romio \
    --with-file-system=nfs+ufs --with-mpe -with-device=ch3:nemesis:tcp,ofi \
    --enable-threads=runtime --enable-fast=all \
    --with-ofi=/uufs/chpc.utah.edu/sys/installdir/libfabric/1.6.0
```

Then start SLURM interactive job with 2 tasks, 2 nodes:

```
$ srun -n 2 -N 2 --reservation=AMDTest -t 4:00:00 -A notchpeak-shared \
        -p notchpeak-shared --pty /bin/bash -l
```

Then compile a hello_mpi code, which prints host name, with the 3.3 and run:

```
notch081]$ /uufs/chpc.utah.edu/sys/installdir/mpich/3.3test/bin/mpiexec -np 2 ./hello_mpi
Hello world from processor notch081, rank 0 out of 2 processors
Hello world from processor notch081, rank 1 out of 2 processors
```

```
notch081]$ /uufs/chpc.utah.edu/sys/installdir/mpich/3.2.1test/bin/mpiexec -np 2 ./hello_mpi
Hello world from processor notch081, rank 0 out of 2 processors
Hello world from processor notch082, rank 1 out of 2 processors
```

Notice, 3.3 ran both tasks on the 1st node, while 3.2.1 correctly distributed them on the 2 nodes.

Running verbosely (printing only select difference in the output), for 3.3:
```
...
host: notch081
...
    Proxy information:
    *********************
      [1] proxy: notch081 (1 cores)
      Exec list: ./hello_mpi (2 processes);
...
Proxy launch args: /uufs/chpc.utah.edu/sys/installdir/mpich/3.3test/bin/hydra_pmi_proxy --control-port notch081:33111 --debug
--rmk user --launcher slurm --demux poll --pgid 0 --retries 10 --usize -2 --proxy-id
```

while for 3.2.1:
```
...
host: notch081
host: notch082
...
    Proxy information:
    *********************
      [1] proxy: notch081 (1 cores)
      Exec list: ./hello_mpi (1 processes);

      [2] proxy: notch082 (1 cores)
      Exec list: ./hello_mpi (1 processes);
...
Proxy launch args: /uufs/chpc.utah.edu/sys/installdir/mpich/3.2.1test/bin/hydra_pmi_proxy --control-port notch081:36883 --debug
--rmk slurm --launcher slurm --demux poll --pgid 0 --retries 10 --usize -2 --proxy-id
```

So, what strikes out that the 3.2.1 has 2 proxies with launch arguments `-rmk slurm --launcher slurm`, while 3.3 only one proxy and launch argument `--rmk user --launcher slurm`.

Which then correlates with srun command as seen through ps, in case of 3.3:

```
/uufs/notchpeak.peaks/sys/installdir/slurm/std/bin/srun --nodelist notch081 -N 1 -n 1 --input none /uufs/chpc.utah.edu/sys/installdir/mpich/3.3i19/bin/hydra_pmi_proxy --control-port notch081:39354 --rmk user --launcher slurm --demux poll --pgid 0 --retries 10 --usize -2 --proxy-id -1
```

while for 3.2.1:
```
/uufs/notchpeak.peaks/sys/installdir/slurm/std/bin/srun -N 2 -n 2 --input none /uufs/chpc.utah.edu/sys/installdir/mpich/3.2.1i18/bin/hydra_pmi_proxy --control-port notch081:3609 --rmk slurm --launcher slurm --demux poll --pgid 0 --retries 10 --usize -2 --proxy-id -1
```

The two differences here are that 3.3 has `--nodelist notch081 -N 1 -n 1` while 3.2.1 has `-N 2 -n 2` in the srun, and `hydra_pmi_proxy` in 3.3 has `-rmk user --launcher slurm` while in 3.2.1 it has `--rmk slurm --launcher slurm`.

Now, if I build the 3.3 with `--with-slurm`, I get the desired effect, that is, the two proxies and srun command as:
```
/uufs/notchpeak.peaks/sys/installdir/slurm/std/bin/srun -N 2 -n 2 --input none /uufs/chpc.utah.edu/sys/installdir/mpich/3.3slurm/bin/hydra_pmi_proxy --control-port notch081:36419 --rmk slurm --launcher slurm --demux poll --pgid 0 --retries 10 --usize -2 --proxy-id -1
```

**Comment by pavanbalaji** [comment]
@mcuma Once you allocate a partition with:

```
$ srun -n 2 -N 2 --reservation=AMDTest -t 4:00:00 -A notchpeak-shared -p notchpeak-shared --pty /bin/bash -l
```

Can you print the following environment variables?  `SLURM_NODELIST`, `SLURM_NNODES` and `SLURM_TASKS_PER_NODE`?

**Comment by mcuma** [comment]
Here it is, with a few more ...
$ env |grep SLURM |grep NODE
SLURM_NODELIST=notch[081-082]
SLURMD_NODENAME=notch081
SLURM_NNODES=2
SLURM_STEP_NUM_NODES=2
SLURM_LAUNCH_NODE_IPADDR=10.242.68.2
SLURM_TASKS_PER_NODE=1(x2)
SLURM_NODEID=0
SLURM_CPUS_ON_NODE=2
SLURM_JOB_NODELIST=notch[081-082]
SLURM_JOB_CPUS_PER_NODE=2(x2)
SLURM_JOB_NUM_NODES=2
SLURM_STEP_TASKS_PER_NODE=1(x2)
SLURM_STEP_NODELIST=notch[081-082]

**Comment by pavanbalaji** [comment]
I think @gcongiu had made some improvements to the slurm detection in `mpich-3.3` to cover additional cases that weren't covered earlier.  Maybe we screwed up something while doing that?

@gcongiu: thoughts?

**Comment by gcongiu** [comment]
@pavanbalaji my patch was improving detection for nodelist formats that were not detected properly. Moreover it also added support for `libslurm` whenever available, which also detects some fancier configuration. I will look into this to see where the problem is.

**Comment by pavanbalaji** [comment]
@gcongiu The problem here is that Hydra is not detecting the node list correctly.  It seems to think that there is just one node allocated.

**Comment by pavanbalaji** [comment]
I believe the bug is here:

https://github.com/pmodels/mpich/blob/master/src/pm/hydra/tools/bootstrap/external/slurm_query_node_list.c#L95

It's assuming a `alphabets-[number_range]` format.  This installation is using `alphabets[number_range]`.  I think the regex should check for `anything[number_range]`

**Comment by gcongiu** [comment]
@mcuma can you try this patch and see if it solves the problem? https://github.com/pmodels/mpich/pull/3578/files#diff-6d8d56dd3adbdf59e03de8a34aef3abe

**Comment by raffenet** [comment]
No response since Feb, I think we can consider this closed. Note that the patch was included in MPICH 3.3.1.
